tools/pipelines/templates/include-publish-npm-package.yml

```diff
@@ -57,23 +57,28 @@ stages:
       pool: ${{ parameters.pool }}
       publishNonScopedPackages: ${{ parameters.publishNonScopedPackages }}
 
-- stage: publish_npm_internal_dev
-  dependsOn: build
-  displayName: Publish to internal dev feed
-  # This condition should match the condition for publish_npm_internal_build
-  condition: and(succeeded(), eq(variables['testBuild'], false))
-  variables:
-  - group: ado-feeds
-  jobs:
-  - template: include-publish-npm-package-deployment.yml
-    parameters:
-      namespace: ${{ parameters.namespace }}
-      feedUrl: $(ado-feeds-dev) # Comes from the ado-feeds variable group
-      feedKind: internal-dev
-      # Environment should match the condition for publish_npm_internal_build
-      environment: package-build-feed
-      pool: ${{ parameters.pool }}
-      publishNonScopedPackages: ${{ parameters.publishNonScopedPackages }}
+# The ony build workflow we have today that needs to publish packages to the dev feed is the one for
+# the client release group. That's the only one where we generate lists of packages to publish to
+# each feed so the only one that should try to publish to the dev feed. Others are fine publishing only
+# to the build feed because the dev feed upstreams it.
```

Comment:
> That's the only one where we generate lists of packages to publish to each feed

That's not quite true; we create lists for [build-tools](https://github.com/microsoft/FluidFramework/tree/main/build-tools/feeds) and [server](https://github.com/microsoft/FluidFramework/tree/main/server/routerlicious/feeds) as well. Won't this change affect those pipelines?

Reply:
I think if we want to make certain release groups not publish to the dev feed, we should update the list generation so that it doesn't list any packages for that feed.

Reply:
Hmm it would, yeah. I wasn't aware of those two :) let me add them. With 3 instances I think it still makes sense to take this approach, but I'd definitely be more open to alternative thoughts or approaches if you have any.

Reply:
My original goal with the changes was to remove logic from the CI pipeline. The more logic that exists outside the pipeline, the easier it is to test and verify things before the pipeline is even involved. With publishing packages in particular, it should be super clear from the lists alone what will be published to any feed. If a release group isn't to be published to a feed, then we should reflect that in the lists.

I still don't understand what's happening, though. All the release groups (there are only three) have lists set up. Is the problem that the build and dev lists aren't mutually exclusive when they should be?

Reply:
The problem is that this template ends up being used by all our build pipelines, including all the independent packages, and we don't have lists for those. So the pipeline is trying to publish the same single package to the build feed and the dev feed, and we sometimes get errors because ADO correctly detects that an upstream (build, of dev) already has the package that we're trying to publish to dev.

If we had lists for every single release group and package, this would not be an issue, but needing lists for every package feels more wrong than this to me.

Reply:
A more explicit option would be a new parameter to the template that says if the "caller" of the template knows that the thing being built has lists or not, and add the stage conditionally on that; the logic is still in the pipeline though.

Reply:
Since the problem is not the release groups, but the independent packages, then IMO a better change would be to exclude the task from those builds. Varying on "release group" vs. "package" makes more sense to me than a condition for specific release groups, though I get that they may be logically equivalent today. We need this information all the time in the pipeline (see the .releaseGroup file in the root of release groups ☹️ ), so a parameter would probably be OK. Especially if the "unscoped packages" parameter can be removed.

> If we had lists for every single release group and package, this would not be an issue, but needing lists for every package feels more wrong than this to me.

We do have lists for every release group - that was required to get it working. I agree lists of a single package doesn't make sense, but I also don't like that making a change from a private to a public package (for exmaple), you have no way of knowing what pipeline that package will wind up being published in until you publish it (or you read docs that are outdated, or you read the pipeline code and understand it all). That seems ridiculous to me.

Honestly, I think the bigger problem is that we are publishing to a feed that is also upstreaming a feed that we are also publishing to. If we instead had feeds that we publish to and SEPARATE feeds that upstream others, this would be much less of a problem.

Reply:
I'll do this with a parameter then to get rid of the errors in the pipelines for now.

> making a change from a private to a public package (for exmaple), you have no way of knowing what pipeline that package will wind up being published in until you publish it

It's true that it wouldn't be obvious to know beforehand, but it's kind of how things work today already? And I think the situation wasn't that terrible before; single packages only went to the build feed (and if public, npmjs, I think? Unless the logic in the pipeline had something to say about that, and that logic hasn't changed). If that's the default (only to the build feed) and we have special handling for the "unusual" situation of "this is not a single package but a release group, and its packages need to be published to different places", it doesn't feel that bad to me.

> Honestly, I think the bigger problem is that we are publishing to a feed that is also upstreaming a feed that we are also publishing to. If we instead had feeds that we publish to and SEPARATE feeds that upstream others, this would be much less of a problem.

Yeah, this might be a good change to consider. I don't think it's inherently wrong to do what we do today, but things would probably be less prone to errors if we made that distinction in our feeds. cc @Abe27342 in case you have thoughts.